echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/PipelineConfigsPollingJob.groovy

```diff
@@ -0,0 +1,193 @@
+  private Scheduler scheduler
+  private Registry registry
+  private PipelineCache pipelineCache
+
+  @SuppressWarnings("unused")
+  PipelineConfigsPollingJob() {
+    pipelineCache = (PipelineCache) SchedulerBeanDependencies.getBean(PipelineCache)
+    registry = (Registry) SchedulerBeanDependencies.getBean(Registry)
+  }
+
+  PipelineConfigsPollingJob(Registry registry, PipelineCache pipelineCache) {
+    this.pipelineCache = pipelineCache
+    this.registry = registry
+  }
+
+  void execute(JobExecutionContext context) {
+    long start = System.currentTimeMillis()
+
+    try {
+      log.info("Running the pipeline configs polling agent...")
+
+      timeZoneId = TimeZone.getTimeZone(context.mergedJobDataMap.getString("timeZoneId"))
+      scheduler = context.scheduler
+
+      // Only interested in pipelines that have enabled CRON triggers
+      def pipelinesWithCronTriggers = pipelineCache.getPipelinesSync().findAll { pipeline ->
+        !pipeline.disabled && pipeline.triggers
+      }
+
+      def pipelineTriggers = new TriggerRepository(pipelinesWithCronTriggers)
+
+      log.debug("Found ${pipelineTriggers.triggers().size()} pipeline CRON triggers that are active")
+
+      removeStaleTriggers(pipelineTriggers)
+      updateChangedTriggers(pipelineTriggers)
+
+      registry.gauge("trigger.count").set(pipelinesWithCronTriggers.size())
+    } catch (Exception e) {
+      log.error("Failed to synchronize pipeline triggers", e)
+    } finally {
+      long elapsedMillis = System.currentTimeMillis() - start
+      log.info("Done polling for pipeline configs in ${elapsedMillis/1000}s")
+      registry.timer("pipelineConfigsPollingAgent.executionTimeMillis").record(elapsedMillis, TimeUnit.MILLISECONDS)
+    }
+  }
+
+  /**
+   * Remove all triggers no longer needed
+   * e.g. triggers that are still registered in our scheduler but no longer found on any pipeline
+   * @param pipelineTriggers all active pipeline CRON triggers
+   */
+  void removeStaleTriggers(TriggerRepository pipelineTriggers) {
+    int removeCount = 0
+    int failCount = 0
+
+    try {
+      Set<TriggerKey> triggerKeys = scheduler.getTriggerKeys(
+        GroupMatcher.triggerGroupStartsWith(PIPELINE_TRIGGER_GROUP_PREFIX))
+
+      triggerKeys.each { triggerKey ->
+        if (pipelineTriggers.getTrigger(triggerKey.name) == null) {
+          try {
+            scheduler.unscheduleJob(triggerKey)
+            removeCount++
+          } catch (Exception e) {
+            log.error("Failed to unschedule job with triggerId: ${triggerKey.name}", e)
+            failCount++
+          }
+        }
+      }
+    }
+    catch (Exception e) {
+      log.error("Failed during stale trigger removal", e)
+    }
+
+    if ((removeCount + failCount) > 0) {
+      log.debug("Removed $removeCount stale triggers successfully, $failCount removals failed")
+    }
+
+    registry.gauge("trigger.removeCount").set(removeCount)
+    registry.gauge("trigger.removeFailCount").set(failCount)
```

Comment:
I think we might want to rethink these metric names, something a little more descriptive. For example: `trigger.sync.removeCount` etc

Reply:
that makes sense, thanks, updating!